internal/restic/find.go

```diff
@@ -1,39 +1,103 @@
 package restic
 
-import "context"
+import (
+	"context"
+	"sync"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// TreeLoader loads a tree from a repository.
+type TreeLoader interface {
+	LoadTree(context.Context, ID) (*Tree, error)
+}
+
+const numUsedBlobsWorkers = 4
+
+// queue implements a concurrency-safe queue of IDs
+// It can be accessed by the channel given with queue.Items()
+type queue struct {
+	q chan ID
+	sync.WaitGroup
+}
```

Comment:
Are you sure the WaitGroup is needed? I may be mistaken, but it seems to me like the following would suffice (untested):

```
type queue struct {
	q    chan ID
	size int64
}

func (q *queue) add(ids IDs) {
	atomic.AddInt64(&q.size, int64(len(ids)))
	go func() {
		for _, id := range ids {
			q.q <- id
		}
	}()
}

func (q *queue) pop() (id ID, ok bool) {
	id, ok = <-q.q
	if ok && atomic.AddInt64(&q.size, -1) == 0 {
		close(q.q)
	}
	return id, ok
}
```

This needs neither Wait nor Done, since FindUsedBlobs is already waiting on an errgroup.Group.

Reply:
From https://golang.org/pkg/sync/atomic/:

> On ARM, x86-32, and 32-bit MIPS, it is the caller's responsibility to arrange for 64-bit alignment of 64-bit words accessed atomically. The first word in a variable or in an allocated struct, array, or slice can be relied upon to be 64-bit aligned. 

I  do not know why the implementation of `WaitGroup` does not use the fact imposed by the last sentence. I changed the code and used it. Can anyone double-check if I understood things correctly?

Reply:
From what I can tell, the alignment guarantees were [more lax](https://github.com/golang/go/commit/fea7c43ea238d8e18f3bafaac2740b3fa8975ae5) when the current [WaitGroup](https://github.com/golang/go/commit/03a48ebe1c79ed0437dcdbc8bf3cd58e901a999d) implementation was written. Then there's the noCopy, but I guess they could move that further down the struct.